src/array/cpu/spmm_blocking_libxsmm.h

```diff
@@ -0,0 +1,598 @@
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_ADD;
+  } else if (std::is_same<Op, op::Sub<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_SUB;
+  } else if (std::is_same<Op, op::Mul<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_MUL;
+  } else if (std::is_same<Op, op::Div<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_DIV;
+  } else if (std::is_same<Op, op::CopyLhs<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_COPY;
+  } else if (std::is_same<Op, op::CopyRhs<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_COPY;
+  }
+  // Second, set which of lhs or rhs is considered first and second operand.
+  // This is needed since libxsmm assumes that the copy operation always copies the first operand.
+  // So, if we need to copy rhs, we need to set that as the first operand.
+  // For rhs, we also set whether to use implicit indices or provided indices.
+  if (std::is_same<Op, op::CopyLhs<DType>>::value) {
+    opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                     LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OPORDER_VECIDX_VECIN);
+  } else if (std::is_same<Op, op::CopyRhs<DType>>::value) {
+    opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                     LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OPORDER_VECIN_VECIDX);
+    if (!has_idx) {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                       LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_IMPLICIT_INDEXED_VECIDX);
+    }
+  } else {
+    opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                     LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OPORDER_VECIDX_VECIN);
+    if (has_idx) {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                       LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_INDEXED_VEC);
+    } else {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                       LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_IMPLICIT_INDEXED_VEC);
+    }
+  }
+  // Third, we set the Redop in the opredop_flags
+  opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | redop_flag);
+  // Fourth, in case of Cmp Redop, set whether to record argmax/argmin for lhs/rhs
+  if (is_cmp) {
+    if (Op::use_lhs) {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                       LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_RECORD_ARGOP_OFF_VEC_0);
+    }
+    if (Op::use_rhs) {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                       LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_RECORD_ARGOP_OFF_VEC_1);
+    }
+  }
+  libxsmm_meltwfunction_opreduce_vecs_idx kernel = nullptr;
+  if (std::is_same<DType, uint16_t>::value) {
+    kernel = libxsmm_dispatch_meltw_opreduce_vecs_idx(
+               N, &_ld, &_ld, LIBXSMM_DATATYPE_BF16, LIBXSMM_DATATYPE_BF16,
```

Comment:
Is this for half precision? If so, we haven't officially supported it yet for CPU (see https://github.com/dmlc/dgl/blob/195f99362d883f8b6d131b70a7868a537e55b786/src/array/cpu/spmm_binary_ops.h#L149). You could safely delete this branch. However, we do allow double as DType and I don't see the branch to handle that.

Reply:
Yes, this is to support BF16. Libxsmm supports BF16 and FP32 so far. It does not support "double" yet. So, I have excluded that path in no_libxsmm condition here:
 https://github.com/dmlc/dgl/pull/3024/files#diff-7f69c0c94acb191f7e1fa6482228a6cf7cf4c71dd69358e142e4d72d5e30ed70R139
We are working on adding support for double in libxsmm but it will not be ready for another two weeks.

Reply:
I have removed support for BF16 for now.